src/test/scala/chiselTests/PrefixSpec.scala

```diff
@@ -0,0 +1,222 @@
+      np_inst.in := in
+    }
+
+    val chirrtl = emitCHIRRTL(new Top)
+
+    val lines = """
+      module Foo_AddOne :
+      module Bar_AddOne :
+      module AddOne :
+      public module Top :
+        inst np_inst of AddOne
+        inst foo_inst of Foo_AddOne
+        inst bar_inst of Bar_AddOne
+        """.linesIterator.map(_.trim).toSeq
+
+    matchesAndOmits(chirrtl)(lines: _*)("AddOne_1")
+  }
+
+  it should "Instantiate should reference the same module definitions when instantiated with the same prefix" in {
+    class Top extends Module {
+      val width = 8
+      val in = IO(Input(UInt(width.W)))
+      val out = IO(Output(UInt(width.W)))
+      val foo_inst1 = withModulePrefix("Foo") {
+        Instantiate(new AddOne(width))
+      }
+
+      val foo_inst2 = withModulePrefix("Foo") {
+        Instantiate(new AddOne(width))
+      }
+
+      foo_inst1.in := in
+      foo_inst2.in := in
+      out := foo_inst1.out
+    }
+
+    val chirrtl = emitCHIRRTL(new Top)
+
+    val lines = """
+      module Foo_AddOne :
+      public module Top :
+        inst foo_inst1 of Foo_AddOne
+        inst foo_inst2 of Foo_AddOne
+        """.linesIterator.map(_.trim).toSeq
+
+    matchesAndOmits(chirrtl)(lines: _*)("AddOne_1", "Bar_AddOne")
+  }
+
+  it should "Memories work" in {
+    class Top extends Module {
+      val io = IO(new Bundle {
+        val enable = Input(Bool())
+        val write = Input(Bool())
+        val addr = Input(UInt(10.W))
+        val dataIn = Input(UInt(8.W))
+        val dataOut = Output(UInt(8.W))
+      })
+
+      val smem = withModulePrefix("Foo") {
+        SyncReadMem(1024, UInt(8.W))
+      }
+
+      val cmem = withModulePrefix("Bar") {
+        Mem(1024, UInt(8.W))
+      }
+
+      val sram = withModulePrefix("Baz") {
+        SRAM(1024, UInt(8.W), 1, 1, 0)
+      }
+
+      smem.write(io.addr, io.dataIn)
+      io.dataOut := smem.read(io.addr, io.enable)
+    }
+
+//    val sv = ChiselStage.emitSystemVerilog(new Top)
+//    println(sv)
+
+    val chirrtl = emitCHIRRTL(new Top)
+    println(chirrtl)
+
+    val lines = """
+      module mem_1024x8(
+      module Top(
+        """.linesIterator.map(_.trim).toSeq
+
+//    matchesAndOmits(sv)(lines: _*)()
+  }
+
+  it should "Definitions that appear within withModulePrefix get prefixed" in {
+    class Top extends Module {
+      val dfn = withModulePrefix("Foo") {
+        Definition(new AddOne(8))
+      }
+
+      val addone = Instance(dfn)
+    }
+
+    val chirrtl = emitCHIRRTL(new Top)
+//    println(chirrtl)
+
+    val lines = """
+  module Foo_AddOne
+  module Top
+        """.linesIterator.map(_.trim).toSeq
+
+    matchesAndOmits(chirrtl)(lines: _*)()
+  }
+
+  it should "withModulePrefix does not affect ExtModules" in {
+    class Sub extends ExtModule {
+    }
+
+    class Top extends Module {
+      val sub_foo = withModulePrefix("Foo") { Module(new Sub) }
+    }
+
+    val chirrtl = emitCHIRRTL(new Top)
+    //println(chirrtl)
```

Comment:
This is [obviously] missing the actual test so make sure to add it and delete any commented out code.